cluster-autoscaler/debuggingsnapshot/debugging_snapshotter.go

```diff
@@ -0,0 +1,222 @@
+	START_DATA_COLLECTION
+	// DATA_COLLECTED is set by setter func (also used by setter func for data collection)
+	// This is set to let Flush know that at least some data collected and there isn't
+	// an error State leading to no data collection
+	DATA_COLLECTED
+)
+
+// DebuggingSnapshotterImpl is the impl for DebuggingSnapshotter
+type DebuggingSnapshotterImpl struct {
+	// State captures the internal state of the snapshotter
+	State *DebuggingSnapshotterState
+	// DebuggingSnapshot is the data bean for the snapshot
+	DebuggingSnapshot DebuggingSnapshot
+	// Mutex is the synchronisation used to the methods/states in the critical section
+	Mutex *sync.Mutex
+	// FlushTrigger is the channel on which the Response Handler waits on to know
+	// when there is data to be flushed back to the channel from the Snapshot
+	FlushTrigger chan bool
+}
+
+// DebuggingSnapshotter is the interface for debugging snapshot
+type DebuggingSnapshotter interface {
+
+	// StartDataCollection will check the State(s) and enable data
+	// collection for the loop if applicable
+	StartDataCollection()
+	// SetNodeGroupInfo is a setter to capture all the NodeInfo
+	SetNodeGroupInfo([]*framework.NodeInfo)
+	// SetUnscheduledPodsCanBeScheduled is a setter for all pods which are unscheduled
+	// but they can be scheduled. i.e. pods which aren't triggering scale-up
+	SetUnscheduledPodsCanBeScheduled([]*v1.Pod)
+	// ResponseHandler is the http response handler to manage incoming requests
+	ResponseHandler(http.ResponseWriter, *http.Request)
+	// IsDataCollectionAllowed checks the internal State of the snapshotter
+	// to find if data can be collected. This can be used before preprocessing
+	// for the snapshot
+	IsDataCollectionAllowed() bool
+	// Flush triggers the flushing of the snapshot
+	Flush()
+	// Cleanup clears the internal data beans of the snapshot, readying for next request
+	Cleanup()
+}
+
+// NewDebuggingSnapshotter returns a new instance of DebuggingSnapshotter
+func NewDebuggingSnapshotter(isDebuggerEnabled bool) DebuggingSnapshotter {
+	state := SNAPSHOTTER_DISABLED
+	if isDebuggerEnabled {
+		klog.Infof("Debugging Snapshot is enabled")
+		state = LISTENING
+	}
+	return &DebuggingSnapshotterImpl{
+		State:             &state,
+		Mutex:             &sync.Mutex{},
+		DebuggingSnapshot: &DebuggingSnapshotImpl{},
+		FlushTrigger:      make(chan bool),
+	}
+}
+
+// ResponseHandler is the impl for request handler
+func (d *DebuggingSnapshotterImpl) ResponseHandler(w http.ResponseWriter, r *http.Request) {
+	// If the debugging snapshot is disabled return 404.
+	// This is a preventive measure in case the handler is listening without the flag
+	// Although in the main we don't listen if the flag is not set
+	if *d.State == SNAPSHOTTER_DISABLED {
+		w.WriteHeader(http.StatusNotFound)
+		return
+	}
+
+	// checks if the handler is in the correct State to accept a new snapshot request
+	if *d.State != LISTENING {
```

Comment:
Checking the state should be done with the mutex locked. Right now there is a race condition between one request checking the state here and another request setting the state back to LISTENING below. In fact, the entire ResponseHandler function would probably be better off in a separate object that interacts with the snapshotter via functions. It would be the handler object then to control whether there are any requests in flight. Snapshotter interface could then have a trigger function accepting a channel that would then be used to return (a pointer to) the result.